File: Selection Sort.java
/**
 * SelectionSort Implementation
 * Time Complexity: Best: O(n^2) | Average: O(n^2) | Worst: O(n^2)
 * Space Complexity: O(1)
 */
public class SelectionSort {
    static int stepCount = 0;

    public static void selectionSort(int[] arr) {
        stepCount = 0;
        int n = arr.length;
        
        for (int i = 0; i < n - 1; i++) {
            int minIndex = i;
            // Find the minimum element in the unsorted array
            for (int j = i + 1; j < n; j++) {
                stepCount++;   
                if (arr[j] < arr[minIndex]) {
                    minIndex = j;
                }
            }
            // Swap the found minimum element with the first element
            if (minIndex != i) {
                int temp = arr[i];
                arr[i] = arr[minIndex];
                arr[minIndex] = temp;
                stepCount++;   
            }
        }
    }

    public static void printArray(int[] arr) {
        for (int num : arr) {
            System.out.print(num + " ");
        }
        System.out.println();
    }

    public static void main(String[] args) {
        int[] arr = {35, 12, 48, 7, 22};
        
        System.out.println("Original Array:");
        printArray(arr);
        
        selectionSort(arr);
        
        System.out.println("Sorted Array (Selection Sort):");
        printArray(arr);
        System.out.println("Step Count: " + stepCount);
    }
}

File: selection_sort.cpp
#include <iostream>
#include <vector>

using namespace std;

int stepCount = 0;

void printArray(const vector<int>& arr) {
    for (int num : arr) {
        cout << num << " ";
    }
    cout << "\n";
}

void selectionSort(vector<int>& arr) {
    stepCount = 0;
    int n = static_cast<int>(arr.size());

    for (int i = 0; i < n - 1; i++) {
        int minIndex = i;

        for (int j = i + 1; j < n; j++) {
            stepCount++;
            if (arr[j] < arr[minIndex]) {
                minIndex = j;
            }
        }

        if (minIndex != i) {
            swap(arr[i], arr[minIndex]);
            stepCount++;
        }
    }
}

int main() {
    vector<int> arr = {35, 12, 48, 7, 22};

    cout << "Original Array:\n";
    printArray(arr);

    selectionSort(arr);

    cout << "Sorted Array (Selection Sort):\n";
    printArray(arr);
    cout << "Step Count: " << stepCount << "\n";

    return 0;
}
